lib/Monad_Lists.thy

```diff
@@ -512,6 +512,33 @@ lemma mapM_wp':
   apply simp
   done
 
+lemma mapM_gets_the_wp:
+  "\<lbrace>\<lambda>s. \<forall>vs. map (\<lambda>t. f t s) ts = map Some vs \<longrightarrow> P vs s\<rbrace>
+   mapM (gets_the \<circ> f) ts
+   \<lbrace>P\<rbrace>"
+unfolding comp_def
```

Comment:
@lsf37 edge case for indentation, where does this one go? it's a new one for me

Reply:
I was going to say this is correct as is, but it is true that we usually put `unfolding` on the same indent as `apply`. That would probably look weird here. What does the Isabelle auto indenter do?

Reply:
The auto indenter indents this by 2 spaces. Sigh. I'll cancel the tests that are running and push an update

Reply:
Feel free to ignore if you don't want to keep on fiddling with this, but I imagine that the unfold isn't needed in the `Nil` case, and if that's right you could just add `comp_def` to the `simp` in the `Cons` case and possibly avoid fighting with the indentation at all.

Reply:
It does seem necessary in both cases, and even adding it to the `simp` in `wpsimp` in both cases doesn't work. So I just went with the `unfolding`. If this is suspicious, I can look into it further

Reply:
Huh, that seems unexpected to me but not enough to be worth further investigation.

Reply:
Yeah. If you want to sweep it under the rug, you can put a `unfold ...,` before the `induct` in `proof`, otherwise yeah just do what auto-indenter does it this case, it's a really rare event it seems.